src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Configuration
+{
+    /// <summary>
+    /// Base class for file based <see cref="ConfigurationProvider"/>.
+    /// </summary>
+    public abstract class FileConfigurationProvider : ConfigurationProvider
+    {
+        public FileConfigurationProvider(FileConfigurationSource source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            Source = source;
+        }
+
+        public FileConfigurationSource Source { get; }
+
+        public override void Initialize(IConfigurationRoot root)
+        {
+            if (root == null)
+            {
+                throw new ArgumentNullException(nameof(root));
+            }
+
+            // TODO: aggregate watch?
+            // REVIEW: should providers validate the source here?
+            if (Source.ReloadOnChange && Source.FileProvider != null)
+            {
+                ChangeToken.OnChange(
+                    () => Source.FileProvider.Watch(Source.Path),
+                    () =>
+                    {
+                        Load();
+                        root.RaiseChanged();
```

Comment:
Would it also work to have IConfigurationProvider expose an individual reload token, and have the root's reload token expire when any provider's reload token is expired? That way the IConfiguration wouldn't need extra APIs like RaiseChanged.